eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -199,10 +203,43 @@ class NodeRelay private(nodeParams: NodeParams,
             context.log.warn(s"rejecting trampoline payment reason=$failure")
             rejectPayment(upstream, Some(failure))
             stopping()
+          case None =>
+            if (!triggered && nextPayload.isAsyncPayment && nodeParams.asyncPaymentsTimeout.isDefined) {
+              waitForTrigger(upstream, nextPayload, nextPacket, nodeParams.asyncPaymentsTimeout.get)
+            } else {
+              doSend(upstream, nextPayload, nextPacket)
+            }
+        }
+      case AsyncPaymentTrigger => context.log.debug("received async payment trigger while waiting to receive all incoming payment packets")
+        receiving(htlcs, nextPayload, nextPacket, handler, triggered = true)
+    }
+
+  private def waitForTrigger(upstream: Upstream.Trampoline, nextPayload: IntermediatePayload.NodeRelay.Standard, nextPacket: OnionRoutingPacket, asyncPaymentsTimeout: FiniteDuration): Behavior[Command] = {
```

Comment:
There is another timeout we need to watch out for: the incoming cltv timeout. We don't want to let the incoming channel force-close because of an HTLC-timeout while we're waiting for the receiver to come back online!

So we should also register to the `CurrentBlockHeight` event, and if we're getting too close to the expiry of the incoming HTLCs, we should fail the payment.

Reply:
More details about how close to the expiry we should fail. We really want to avoid seeing force-closed channels with the sender caused by an offline recipient. So we'll probably need to take a very large buffer because once we decide we want to fail the payment, we need to wait for the sender to come back online to actually sign the `update_fail_htlc` (and that needs to happen before the incoming payment expires). We reasonably need to allow at least 144 blocks for the sender to come back online, what do you think?

We should probably add another configurable value in `reference.conf` to control that: `relay.async-payment-cancel-safety-before-timeout-blocks = 144`? Then in the `waitForTrigger` state, at each block we would check if `currentBlockHeight > upstream.expiryIn - asyncPaymentCancelSafetyBeforeTimeout`, and if so, we'd cancel the payment.

Now that we have two configuration values, it makes sense to group them in a dedicated object in `reference.conf`, and create a corresponding case class inside `RelayParams`:

```
relay {
  async-payments {
    timeout = 7 days
    async-payment-cancel-safety-before-timeout-blocks = 144
  }
}
```

When reading the configuration, we should also verify that the `timeout` is at least twice the cancel safety, otherwise the feature will be quite useless.

Reply:
This makes sense to me. This value is similar to `channel.expiry-delta-blocks` for an immediately forwarded payment. If a relay forwards a payment, then it should have claimed or failed the downstream payment somehow (either on or off chain) `channel.expiry-delta-blocks` before the upstream payment expires. For waiting payments that have not been forwarded yet at that same point, then a relay should initiate a payment failure with the same (or more) lead time.

When reading the configuration, should we also fail if `async-payment-cancel-safety-before-timeout-blocks` < `channel.expiry-delta-blocks`?  Looking at NodeRelay.scala, it seems like it would be strictly unsafe to forward a payment after the `payloadOut.outgoingCltv` block which must be at least `nodeParams.channelConf.expiryDelta` sooner than `upstream.expiryIn` according to the check in `validateRelay`.

The `timeout` parameter prevents our node from holding onto many unclaimed payments with really long `upstream.expiryIn` values. Perhaps another way to define `timeout` would be to set a cap on the longest `upstream.expiryIn` we're willing to wait for? That way everything would be in block time. On the other hand, `timeout` is a nice human readable way of configuring how long to hold async-payments.

Reply:
> When reading the configuration, should we also fail if async-payment-cancel-safety-before-timeout-blocks < channel.expiry-delta-blocks? 

Yes, that's a good idea!

>  Perhaps another way to define timeout would be to set a cap on the longest upstream.expiryIn we're willing to wait for?

The issue with that is that we don't have a good mechanism to communicate this to the sender. I don't think it would be necessary to be honest, as it would be a poor UX to go back and forth with the user to tell them to set a smaller delay, that's something that should be hidden from them as much as possible.

I think we should let node operators choose that `timeout` regardless of what `expiryIn` they receive. I don't know if that `timeout` should be in blocks or time, I think both would work? Maybe using blocks is a bit more consistent?

Reply:
I went with replacing `timeout` with `hold-timeout-blocks`. The `NodeRelay` actor will only wait for this many blocks (default: 1008 or ~ 7 days) before timing out the relay.

I also added the `cancel-safety-before-timeout-blocks` parameter and will timeout that many blocks (default: 144)  before an incoming payment is set to expire.

Fixed in [fcdc394](https://github.com/ACINQ/eclair/pull/2435/commits/fcdc394cb2751d7408318afb437cf49d18389d08)